### ISSUE TITLE
Project maintainer wanted

### ISSUE THREAD

**Issue by civctp2** [civctp2/civctp2#2]
Contribution to the project are on hold, until maintainer is found.
For details please refer to this discussion: https://github.com/ptitSeb/ctp2/issues/3#issuecomment-427801575

**Comment by DEVoytas** [comment]
Since @ptitSeb said in this comment: https://github.com/civctp2/civctp2/issues/21#issuecomment-441194286 that we will not be able to act as a temporary maintainer, I propose two alternative solutions (until real maintainer is found):
- either @LynxAbraxas agrees for this role (he crated many very useful PRs so he is most active and motivated person in this project at the moment)
- I will start merging pending PRs, but since I (as stated many times before) do not feel authorized to make such decisions, I would only merge those, that we all agree on (but take into account I may have periods of lower activity)

@LynxAbraxas, @ptitSeb what do you think about it? Do yo have any other ideas?

**Comment by ptitSeb** [comment]
I'm ok with that @DEVoytas .

**Comment by LynxAbraxas** [comment]
Yes, pleas go ahead @DEVoytas! My current New Years resolution for 2019 is to not work on this any more (for at least a year). It has infected me again...

**Comment by LynxAbraxas** [comment]
@MartinGuehmann: Would you mind to be maintainer? You have profound knowledge of civctp2 source and of its evolution.

**Comment by MartinGuehmann** [comment]
For sure, I can do this, I just have to get familiar with GitHub.

So the tasks are merging the branches. We have the two mentioned here and there is also trunk it is 11 commits ahead of master. I don't exactly know what these commits are about.

**Comment by MartinGuehmann** [comment]
> @MartinGuehmann that is great! Now we need to figure out how can I pass `civctp2` account credentials to you. Maybe email?

@DEVoytas, I guess email is fine. You can make GitHub display the email address, publically. But of course in the age of spam-bots, that's not a good idea. I can't make GitHub display it only to some people, so let's do it like this:

guehmann
at
t-online.de

> 
> PS.
> I understand the need of getting familiar with GitHub, but are you more or less familiar with git in general?

I played around with the Git stuff in the last days. So it is more complex than Subvision, but that shouldn't be much of a problem.

The first goal is to bring master into a releasable stage, at least for Windows. Of course master should also compile on Linux, but I haven't managed to do this, yet. It stuck at wldgen.cpp. I haven't checked the compiler. I just used what came with Kubuntu.

Right now, I know that "Now it compile on 64Bits platforms" (Git doesn't have these nice revision numbers) breaks compiling on Windows, but up then everything is fine. This version behaves the same as Revision 1111, if you compile it with VC 2008.  The behavior of the AI changes a little on VC 2017 and a few fixes for compiling. But so far I can assume it is just the different build environment. The same happened with compiling on VC6. My feeling was on VC6 the AI behaved a little bit better. Therefore, I always released VC6 builds.

The current head of master is broken. The AI does not behave not only differently, but it is differently in a broken way. It doesn't explore, and it doesn't improve tiles, properly. The road network is broken and no commerce improvements and also some other things. I assume that these intptr _t, maybe a problem, but there were also some other things, I have seen. But that should be easy to sort out.

More difficult are the changes from the Roland Taverner branch. They don't seem to break the AI, but they make the game crash. And that is quite some code that was changed. We have in simplifying the code, which the new standard allows, probably I take this, because these iterators are quite hard to read. Replacing the raw pointers is a bigger task. And I am not sure whether we should do this. Since the debug version has the leak reporter, we should find the leaks.

**Comment by LynxAbraxas** [comment]
> So the tasks are merging the branches. We have the two mentioned here and there is also trunk it is 11 commits ahead of master. I don't exactly know what these commits are about.

I think a lot of contributions are similar and originate from the fact that they were done out of sync. Some were on SVN trunk, others on non-deterministic git ports. I worked with the contribs of @ptitSeb because they make it possible to compile master on linux and provide a way to avoid FAT installs which is the main reason that ctp2 can now be not only compiled inside a docker environment but also played from a docker image (DI), which I find most appealing (see my other posts for details).

The intptr _t cased inappropriate game behaviour, see e.g. #33. Reverting back to the former code and using `-fpermissive` for compilation seems to solve the problem, see #35 and #36.

All issues and PRs reported by me are for the compilation with SDL under Linux inside the docker environment defined by the Dockerfile in https://github.com/LynxAbraxas/ctp2DF.

With the changes comprised of all my PRs [master4ctp2DF](https://github.com/LynxAbraxas/civctp2/tree/master4ctp2DF) (some debugging, more additional features), the main problem that I face when playing with the DI version of master, are the crashes appearing at random intervals (sometimes within the first minute after game start, sometimes after hours of playing), see #30.

**Comment by MartinGuehmann** [comment]
Originally, I would have gone for the branch of Roland Taverner, but that one is more difficult to fix. It also makes the code compile under Windows with VC 2017, but the project files are a bit ugly, so I ended up to convert my own VC 2008 project files. Basically for VC 2017 compilation, I also did the same changes as in that branch. It seems to work everything except the intro move, just sound but no picture.

Anyway with the fix of #36, the AI seems to be fine, almost like before, just for minor differences. The max template function uses float, even so it's supposed to be double.

**Comment by MartinGuehmann** [comment]
> > For sure, I can do this, I just have to get familiar with GitHub.
> 
> @MartinGuehmann that is great! Now we need to figure out how can I pass `civctp2` account credentials to you. Maybe email?
> 
> PS.
> I understand the need of getting familiar with GitHub, but are you more or less familiar with git in general?

@DEVoytas, I am still waiting. The alternative of sending me the credentials, you could allow me to commit to the civctp2 repository. Something, I would do this way, anyway.

**Comment by DEVoytas** [comment]
Hi @MartinGuehmann, apologies for the delay. I've just sent account details to your email, so please check if you can log in now.

**Comment by civctp2** [comment]
@DEVoytas, obviously I can log in, thanks. I added myself as a collaborator. Let's see how that works out.

**Comment by DEVoytas** [comment]
@civctp2 , @MartinGuehmann great! So I guess we can now merge following PR: #56 and close this issue: https://github.com/civctp2/civctp2/issues/2.

**Comment by MartinGuehmann** [comment]
#56 is merged in, and this issue is now closed.

**Comment by LynxAbraxas** [comment]
Thanks @MartinGuehmann for taking the job as maintainer and thanks @DEVoytas for all the initiatives. It's really great to see the ctp2 source code project revived.

@MartinGuehmann: From the commit history I see you have been busy a lot already. Thanks for that. I would recommend though that you also create PRs for the commits you make and want to integrate in to ctp2@master such that ideally master only contains PR merge commits, as is the recommended git/GH way of contributing. This way it is easier for other contributors to keep track of what is where and the change can be discussed in its own thread (PR). Currently, I cannot find the video-tutorial (from someone working at GH) I once got those ideas from but you can have a look at https://help.github.com/articles/what-is-a-good-git-workflow/ or https://gist.github.com/Chaser324/ce0505fbed06b947d962, perhaps @DEVoytas has some other great links.

**Comment by LynxAbraxas** [comment]
For example: if https://github.com/civctp2/civctp2/commit/6d817f4cc45c1e96b52f7ab666a3137b58912d75 had been in a separate branch for a PR, commit https://github.com/civctp2/civctp2/commit/55ecfc4de02cf8c6d09a5ed9eeb381d56e4b4b83 would probably have followed directly onto it or both would have been squeezed into one and tested by the build system and only merged into master when the builds succeeded.
If you look at https://github.com/civctp2/civctp2/commits/master you can see that e.g.
https://github.com/civctp2/civctp2/commit/55ecfc4de02cf8c6d09a5ed9eeb381d56e4b4b83
https://github.com/civctp2/civctp2/commit/501d56c1c9b0ea8e8430d032bdd17d3336929d1c
https://github.com/civctp2/civctp2/commit/f22821505d20ff5bbb698dc3db9828c145b53f1e
did not compile on the build system (red cross), such commits should not be part of master but be on branch with a PR. 
Contribution rules on some projects even demand that every commit on any branch compiles and passes all tests, although I am a fan (so far) for not squeezing all commits to preserve the info of what did not work/suffice, see e.g. https://github.com/civctp2/civctp2/pull/13#issuecomment-442937860.

**Comment by LynxAbraxas** [comment]
If civctp2-master should stay as is, I will test it on GL with https://github.com/LynxAbraxas/ctp2DF. However it was accidentally pushed I would vote for resetting master to https://github.com/civctp2/civctp2/commit/74c4f072fa185716eef73fd0393689a5112a27ad and then do clean PR merges for the various changes.

**Comment by DEVoytas** [comment]
> Thanks @MartinGuehmann for taking the job as maintainer and thanks @DEVoytas for all the initiatives.

Big thanks to both of you as well @LynxAbraxas and @MartinGuehmann as well!

>  I would recommend though that you also create PRs for the commits you make (create your own fork of civctp2 

I agree, actually already mentioned that here: https://github.com/civctp2/civctp2/issues/57

> This way it is easier for other contributors to keep track of what is where and the change can be discussed in its own thread (PR)

Also agree. Final decision is always up to the maintainer, but it is always good practice to have some discussion and let others give their opinion about all changes that are meant to be merged.

> I would vote for resetting master to 74c4f07 and then do clean PR merges for the various changes.

Normally rewriting the history of a public branch it is very strongly discouraged (see https://stackoverflow.com/questions/11149499/what-happens-when-you-rewrite-history-of-a-public-branch#11149606 for reasons why). However given that not many people has cloned the repo so far (judging by the number of forks on GH) it may actually be worth, it in order to have cleaner history (which does not really look so  good right now).
From network view: https://github.com/civctp2/civctp2/network it looks like no one based their work on most recent master, and there is rather low chance that someone cloned the repo and does some work outside of GH.
So on one hand this is really bad practice and the other benefits seem to outweigh the downsides in this case... I cannot make my mind on this so I guess I will abstain from  giving my opinion,

**Comment by MartinGuehmann** [comment]
> @MartinGuehmann: From the commit history I see you have been busy a lot already. Thanks for that. I would recommend though that you also create PRs for the commits you make (create your own fork of civctp2 at https://github.com/MartinGuehmann, push to that and create PRs to https://github.com/civctp2/civctp2/ that can be discussed before being merged at GH by you or civctp2 user) and want to integrate in to ctp2@master such that ideally master only contains PR merge commits, as is the recommended git/GH way of contributing. This way it is easier for other contributors to keep track of what is where and the change can be discussed in its own thread (PR).

@LynxAbraxas, I created my own fork and pushed the first stuff there. Actually, also fixing my own bug, but well. Right now it is also my master branch, probably also not the best location, but for the start it should be good enough.

**Comment by MartinGuehmann** [comment]
> Contribution rules on some projects even demand that every commit on any branch compiles and passes all tests, although I am a fan (so far) for not squeezing all commits to preserve the info of what did not work/suffice, see e.g. [#13 (comment)](https://github.com/civctp2/civctp2/pull/13#issuecomment-442937860).

I wouldn't squeeze anything, either. It doesn't feel like a good idea to tinker around with it. We should keep in mind that this project has a non-Git history. This means not every commit compiles, anyway. So, I would rather demand that only the head passes all tests. The thing is it is about coding and move things, so I would like to put the bureaucracy to a minimum. Obviously, we need some, but it shouldn't be not more then necessary.

**Comment by MartinGuehmann** [comment]
> @MartinGuehmann Concerning the above, after goning through some of the new commit to master I am beginning to wonder if you accidentally pushed your local testing master to civctp2-master?

Well, it is my local master. This just reflects the subversion workflow. Where you rather download the things from the server, merge them into your working copy and then upload once you are done. We still have issue #1, which is an issue because it was branched. As long the person who feels responsible for the branch is around and cares, we have no problem the branch won't go out of sync. But once, the person disappears, nobody would keep it in sync, and then things go bad. To merge the Linux branch properly, I think I have to do the brute force method: To go through every file and merge manually what I think is was merging. That is quite some work, so for me it is a nightmare to have branches laying around in this project. Of course, it would have been our burden to make the game compile on Windows again, with the Linux changes that may occasionally break things on Windows. But it would have been the minor costs. 

> For example I only created an issues (#30) but never created a PR for [debug-random-segfaults](https://github.com/LynxAbraxas/civctp2/tree/debug-random-segfaults) but it got included into civctp2-master here:
> [fcfefc6](https://github.com/civctp2/civctp2/commit/fcfefc61a5480104c3a9e1f400c8547dca350705)
> Which is not good because it contains commits from other PRs that I needed to include to test if they were the origin for the random segfaults. Anyway this branch as a PR would be WIP and not ready for merging.
> 
> How did you stumble up on [debug-random-segfaults](https://github.com/LynxAbraxas/civctp2/tree/debug-random-segfaults)?

As with the things that should have been merged in, I fetched them from the server.  With all these pull requests lying around it got a bit confusing. I couldn't merge them on GitHub, either. The debug-random-segfaults is one file to revert, that should be easy to do, because as far as I can see only one file came from it. Now, I have the following in master:

DEVoytas/pm_readme_update #56
LynxAbraxas/activateAllGoods #44
LynxAbraxas/spriteCreation #46
LynxAbraxas/spriteEditor4Goods #45
LynxAbraxas/spriteEditor #41
LynxAbraxas/debugKeyHandling #40
LynxAbraxas/goodMushrooms #47
LynxAbraxas/debug-random-segfault
LynxAbraxas/compile4debug #39 x
LynxAbraxas/revert-intptr_t #36
LynxAbraxas/debugDiplomacy #35
LynxAbraxas/SDLvideo4master #34
LynxAbraxas/mouse-wheel #26
LynxAbraxas/goodsAnimDebug #24
LynxAbraxas/parallelBuild #17
LynxAbraxas/debugLoadGame #18

PR #39 did not build, but I fixed it. And then I have only debug-random-segfault to remove. What I can already say about is that the problem is not there. But it is something that has been fixed in the Linux branch.

On the Linux branch I have:
LynxAbraxas:linux-segfault-debug #25
LynxAbraxas:docker4linux #37

docker4linux already contained LynxAbraxas:linux4ctpDF. So to have these things should be fine, I added on top the VS files for browsing the project on Windows. I assume this is also uncontroversial.

Then I also updated the VS project files on the RolandTaverner's branch to make the game compile according to the manual.

**Comment by MartinGuehmann** [comment]
Obviously, we have to talk about something, so I reopened the issue. Ideally, I would split this part off, but I don't know how.

> If civctp2-master should stay as is, I will test it on GL with https://github.com/LynxAbraxas/ctp2DF.
> 
> However it was accidentally pushed I would vote for resetting master to [74c4f07](https://github.com/civctp2/civctp2/commit/74c4f072fa185716eef73fd0393689a5112a27ad) and then do clean PR merges for the various changes. I do realize that this would be some work considering how busy @MartinGuehmann was ;-) but probably would be worth the effort due to the fact that some things got included that should not be and some multiple times and currently it is very hard to follow through the late history chaos.

Well, if you wanna do this. ;) For removing what nobody has requested to include, I would just need to change ../ctp2_code/ui/aui_sdl/aui_sdlblitter.cpp. And that's it. As far as I understand it turns segfault into deadlocks. That's exactly what I get. I prefer the segfaults this way I don't have to kill CTP2 myself. But whether deadlocks or segfault, either way CTP2 for Linux from master isn't in a good state.

> [74c4f07](https://github.com/civctp2/civctp2/commit/74c4f072fa185716eef73fd0393689a5112a27ad) was the latest commit when I last fetched https://github.com/civctp2/civctp2 at the beginning of the year.
> It seems some of my branches/PRs got merged outside GH such that I got no message about the new status and therefore could not comment on new commits, nor did I get informed by all the great work that was done by @MartinGuehmann lately. Please merge on GH or others wont be able to follow.

I think, I just remove what should not go into master and agree on the rules for merging. And agree on the rules to adhere for the future.

No direct push to master.
The master branch is only updated via pull requests.
Pull requests can only be merged in if they pass Travis-CI.
Any head commit should build on Linux, ideally also on Windows but Travis-CI does not check this right now.
No, history flattening, unless you still request to do it for the current master. But that would be the only exception.

For the pull requests, I would like to merge them in as soon as possible, at max a week, unless there is more work to do on them. So what would be the minimum time to keep them open, I guess at least a day. Or should I merge them in as soon as they have passed the checks?

**Comment by LynxAbraxas** [comment]
> I wouldn't squeeze anything, either. It doesn't feel like a good idea to tinker around with it. We should keep in mind that this project has a non-Git history. This means not every commit compiles, anyway. So, I would rather demand that only the head passes all tests.

My preferred way for my own projects is to tell the history of trail-and-error in side-branches (only squashing/reordering commits if I forgot something or made a mistake) allowing commits not passing tests, but being strict that any commit to master does, ideally master only to be made up of merge commits that used `--no-ff`. If you then look at the git history with e.g. `gitk` you can get the the squeezed history on the short branch side and the full story on the long. With a right click on another commit you can even get the diff over the range of commits.

**Comment by LynxAbraxas** [comment]
For commits like https://github.com/civctp2/civctp2/commit/19fc4acf7bc1e50c390a3f14cfd71b77fc692c91 it is best to use `git cherry-pick`. Even better would be to create a new topic-branch on master, cherry-pick to that, then create a PR.

**Comment by MartinGuehmann** [comment]
> Concerning resetting master: From looking at the current history, there is more in from me than there should, but I agree that reverting the changes from https://github.com/LynxAbraxas/civctp2/tree/debug-random-segfaults is probably the shorter option. I leave it to @MartinGuehmann as maintainer to decide.

Actually, since I did not understand the acronym WIP (I commented in #57 about it), I am more seriously thinking about. But still it is some work, it may be less work than I thought, initially. But it is still something I should not decide on so easily.

> With #63 compilation with SDL_ffmpeg on GitLab works again but not all test of https://github.com/LynxAbraxas/ctp2DF succeed.

Well, that are regressions, which would have to be resolved anyway, whether I would rewrite history or not.

**Comment by LynxAbraxas** [comment]
> Well, that are regressions, which would have to be resolved anyway, whether I would rewrite history or not.

With git, this can be avoided when carefully doing the merge commit. Git allows to have two or multiple way merges (with special logic) which combine changes from both branches that are combined, you can even add further changes (overlooked first) with `git commit --amend` to a merge commit. Git's automatic merging mostly decides well which changes have precedence over other others, but not in all cases. It would have been nice to have the "regression" of #63 be part of the merge commit that introduced it. With the order of merging the PRs into master as in https://github.com/LynxAbraxas/civctp2/commits/master4ctp2DF?after=20eb4b309803f2d06fa6c09b658ed0e554d53e33+69 git's precedence decision avoided such a regression.

**Comment by DEVoytas** [comment]
@LynxAbraxas @MartinGuehmann the discussion here seem to be no more related to finding project maintainer. I suggest closing the issue and continue it in dedicated issue I created here: https://github.com/civctp2/civctp2/issues/65
Please read it and see if you agree.
If there is any specific information it this issue that should not be lost/overlooked, it is very easy to link to specific comment or quote it.

**Comment by MartinGuehmann** [comment]
I haven't figured out how to link to a specific post, but let's do as @DEVoytas suggest not here.
